File: ComponentModel/Graphics/LineRenderer.cs
using System.Collections.Generic;
using System.Xml;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using FPX;
using System;

public class LineRenderer : Component, IDrawable
{
    private const float UpdateInterval = 0.1f;
    private float updateTimer = UpdateInterval;

    public bool useWorldSpace = true;

    public List<Vector3> positions = new List<Vector3>();

    public List<VertexPositionColor> vertecies = new List<VertexPositionColor>();

    public BasicEffect effect;

    public Material material;

    public event EventHandler<EventArgs> DrawOrderChanged;
    public event EventHandler<EventArgs> VisibleChanged;

    public int DrawOrder { get; set; }

    public bool Visible { get; set; } = true;


    public void Start()
    {
        if (material == null)
            material = gameObject.AddComponent<Material>();

        for (int i = 0; i < positions.Count; i++)
            vertecies.Add(new VertexPositionColor(positions[i], Color.White));

        effect = new BasicEffect(GameCore.graphicsDevice);
    }

    public void Update(GameTime gameTime)
    {
        updateTimer -= Time.deltaTime;
        if (updateTimer < 0.0f)
        {
            for (int i = 0; i < positions.Count; i++)
            {
                if (i < vertecies.Count)
                {
                    var vertex = vertecies[i];
                    vertex.Position = positions[i];
                    vertex.Color = material.DiffuseColor;
                    vertecies[i] = vertex;
                }
                else
                    vertecies.Add(new VertexPositionColor(positions[i], material.DiffuseColor));
            }

            updateTimer = UpdateInterval;
        }
    }

    public void Draw(GameTime gameTime)
    {
        var bs = GameCore.graphicsDevice.BlendState;
        GameCore.graphicsDevice.BlendState = material.blendState;

        effect.View = Camera.Active.ViewMatrix;
        effect.Projection = Camera.Active.ProjectionMatrix;
        if (useWorldSpace)
            effect.World = transform.worldPose;

        effect.DiffuseColor = material.DiffuseColor.ToVector3();
        effect.SpecularColor = material.SpecularColor.ToVector3();
        effect.CurrentTechnique.Passes[0].Apply();

        GameCore.graphicsDevice.DrawUserPrimitives(PrimitiveType.LineStrip, vertecies.ToArray(), 0, vertecies.Count - 1);
        GameCore.graphicsDevice.BlendState = bs;
    }

    public override void LoadXml(XmlElement element)
    {
        var positionsNode = element.SelectSingleNode("Positions");
        var useWorldSpaceNode = element.SelectSingleNode("UseWorldSpace");

        if (positionsNode != null)
        {
            foreach (XmlElement node in positionsNode.ChildNodes)
            {
                Vector3 pos = LinearAlgebraUtil.Vector3FromXml(node);
                positions.Add(pos);
            }
        }

        if (useWorldSpaceNode != null)
            useWorldSpace = bool.Parse(useWorldSpaceNode.InnerText);
    }
}

File: ScriptEngine/LineRenderer.vb
Imports System.Xml
Imports Microsoft.Xna.Framework
Imports Microsoft.Xna.Framework.Graphics
Imports FPX

Public Class LineRenderer
    Inherits Component

    Public useWorldSpace As Boolean = True

    Public positions As List(Of Vector3) = New List(Of Vector3)

    Public vertecies As List(Of VertexPositionColor) = New List(Of VertexPositionColor)()

    Public effect As BasicEffect

    Public Sub Start()
        For i = 0 To positions.Count - 1
            vertecies.Add(New VertexPositionColor(positions(i), GetComponent(Of Material).DiffuseColor))
        Next

        effect = New BasicEffect(GameCore.graphicsDevice)
    End Sub

    Public Sub Draw(gametime As GameTime)
        effect.View = Camera.Active.ViewMatrix
        effect.Projection = Camera.Active.ProjectionMatrix
        If Not useWorldSpace Then
            effect.World = transform.worldPose
        End If

        effect.DiffuseColor = GetComponent(Of Material).DiffuseColor.ToVector3()
        effect.SpecularColor = GetComponent(Of Material).SpecularColor.ToVector3()
        effect.CurrentTechnique.Passes(0).Apply()

        GameCore.graphicsDevice.DrawUserPrimitives(Microsoft.Xna.Framework.Graphics.PrimitiveType.LineList, vertecies.ToArray(), 0, vertecies.Count / 2)
    End Sub

    Public Overrides Sub LoadXml(element As XmlElement)
        Dim positionsNode = element.SelectSingleNode("Positions")
        Dim useWorldSpaceNode = element.SelectSingleNode("UseWorldSpace")

        If Not positionsNode Is Nothing Then
            For Each node In positionsNode.ChildNodes.Cast(Of XmlElement)
                Dim pos = LinearAlgebraUtil.Vector3FromXml(node)
                positions.Add(pos)
            Next
        End If

        If Not useWorldSpaceNode Is Nothing Then
            useWorldSpace = Boolean.Parse(useWorldSpaceNode.InnerText)
        End If
    End Sub

End Class
